server/rest/src/main/java/org/infinispan/rest/stream/CacheChunkedStream.java

```diff
@@ -0,0 +1,101 @@
+package org.infinispan.rest.stream;
+
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.stream.ChunkedInput;
+import io.netty.handler.stream.ChunkedStream;
+
+public abstract class CacheChunkedStream<T> extends InputStream implements ChunkedInput<ByteBuf> {
+
+   protected InternalStreamConsumer<T> consumer;
+   private final ChunkedStream delegate;
+   private volatile  boolean closed;
+
+   public CacheChunkedStream() {
+      this.delegate = new ChunkedStream(this);
+   }
+
+   @Override
+   public boolean isEndOfInput() {
+      return closed;
+   }
+
+   protected final void flushedAll() {
+      closed = true;
+   }
+
+   @Override
+   public void close() {
+      try {
+         delegate.close();
+      } catch (Exception e) {
+         throw new RuntimeException(e);
+      }
+   }
+
+   @Override
+   public ByteBuf readChunk(ChannelHandlerContext ctx) throws Exception {
+      return delegate.readChunk(ctx);
+   }
+
+   @Override
+   public ByteBuf readChunk(ByteBufAllocator allocator) throws Exception {
+      return delegate.readChunk(allocator);
+   }
+
+   @Override
+   public long length() {
+      return delegate.length();
+   }
+
+   @Override
+   public long progress() {
+      return delegate.progress();
+   }
+
+   protected static class InternalStreamConsumer<T> implements Runnable, Iterator<T> {
+      private final BlockingQueue<T> queue;
+      private final Iterator<? extends T> iterator;
+      private T next;
+      private volatile boolean hasNext;
+
+      protected InternalStreamConsumer(Iterator<? extends T> iterator) {
+         this.queue = new ArrayBlockingQueue<>(64);
+         this.iterator = iterator;
+         this.next = iterator.next();
+      }
+
+
+
+      @Override
+      public void run() {
+         while (iterator.hasNext()) {
+            try {
+               queue.put(iterator.next());
+            } catch (InterruptedException e) {
+               Thread.currentThread().interrupt();
+               break;
+            }
+         }
+         hasNext = false;
+      }
+
+      @Override
+      public boolean hasNext() {
+         return next != null || hasNext || !queue.isEmpty();
+      }
+
+      @Override
+      public T next() {
+         T current = next;
+         next = queue.poll();
```

Comment:
This uses a lock internally, is that a problem? 

I would expect only a single consumer here, but I could be wrong, as I am not sure how Netty will read the stream.

Reply:
This should be fine, you can add an exception to block hound integration class as needed. We do this for similar methods in other places.